### ISSUE TITLE
Doc updates for Epic 20168

### ISSUE THREAD

**Issue by teddyjtorres** [OpenLiberty/docs#5646]
Tentative release is 22.0.0.9.

For Epic https://github.com/OpenLiberty/open-liberty/issues/20168, we need to add the following to the https://openliberty.io/docs/latest/reference/feature/openidConnectClient-1.0.html#jwt section, preferably at the end (just before "Claim to subject mapping"),

Open Liberty allows using the issuer claim from a JWT or JWS access token to select the `openidConnectClient` configuration with a matching `issuer` attribute. For an issuer claim value of `"https://hostname/op1"` and a server.xml file configured with,

```
<openidConnectClient id="RP1" issuer="https://hostname/op1"  ... />
<openidConnectClient id="RP2" issuer="https://hostname/op2" authFilterRef="rp2filter" .../>

<authFilter id="rp2filter">
  ...
</authFilter>
```
the RP1 configuration is selected as long as the "rp2filter" does not apply to the request. Therefore, an attempt to find which openidConnectClient applies to the request will be done using the following algorithm:

1. If the auth filter for a openidConnectClient matches the request, choose that configuration.
2. Check the issuer from the JWT against the issuers of all config elements. If the issuer from the JWT matches a single openidConnectClient configuration element, choose that configuration.
3. If the issuer from the JWT matches the issuer of multiple openidConnectClient configuration elements, choose the first configuration matched.
4. If the issuer does not match the issuer of any configuration elements, choose between all of the config elements in a non-deterministic manner.

**Comment by ramkumar-k-9286** [comment]
Hi Teddy @teddyjtorres 

Suggested changes have been made.

Draft Doc Link: https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/reference/feature/openidConnectClient-1.0.html#jwt

Please check the same and let me know if any additional changes are to be made to the documents.

Please add the `Developer Reviewed` label if the changes made are to your satisfaction.

Regards,
Ramkumar

CC @dmuelle

**Comment by teddyjtorres** [comment]
Thank you @ramkumar-k-9286. The changes are good and I added the "Developer reviewed" label.

**Comment by dmuelle** [comment]
# Peer review

> Open Liberty allows by using the issuer claim from a JWT or JWS access token to select the openidConnectClient configuration with a matching issuer attribute.
--->
> Open Liberty can use the issuer claim from a JWT or JSON web signature (JWS) access token to select a `openidConnectClient` configuration in the `server.xml` file with a matching issuer attribute.

---
> An issuer claim value of "https://hostname/op1" and a server.xml file that is configured with the RP1 configuration is selected if the "rp2filter" does not apply to the request.
--->
> The following `server.xml` file example shows two Open ID connect client configurations, with an **authentication filter** (link) configured to route requests to the `RP2` configuration. If a request with an issuer claim value of `https://hostname/op1` is presented, the `RP1` configuration is selected unless the `rp2filter ` applies to the request.


I suggest the following revisions to make the algorithm flow clearer.  I think you should restore the numbered steps to the algorithm since OL presumably follows that order/priority.


Open Liberty selects the `openidConnectClient` configuration to use for a JWT request according to the following algorithm.

1. If the authentication filter for a `openidConnectClient` configuration matches the request, choose that configuration.
2. If the issuer claim from the JWT matches the `issuer` attribute in only one `openidConnectClient` configuration element, choose that configuration.
3. If the issuer claim from the JWT matches the `issuer` attribute in multiple `openidConnectClient` configuration elements, choose the first configuration that matches.
4. If the issuer claim from the JWT does not match the `issuer` attribute in any `openidConnectClient` configuration elements, choose between all of the configuration elements in a nondeterministic manner.

I think we need to be more clear about the results of (4)- what happens in this case? It chooses a config element basically at random? Does it attempt top validate the request on criteria other than the issuer claim? If it doesn't find any matching criteria in the first non-deterministically chosen config, does it keep checking through any others until all options are exhausted? Or does the request fail somehow?

**Comment by ramkumar-k-9286** [comment]
**Communication with Teddy on Slack**

**Ramkumar K [1:57 PM]**
Hi Teddy
As a part of the Peer Review - David added a comment on the issue.
Could you provide some clarity in that case?  Question added
4. If the issuer does not match the issuer of any configuration elements, choose between all of the config elements in a non-deterministic manner.
I think we need to be more clear about the results of (4)- what happens in this case? It chooses a config element basically at random? Does it attempt top validate the request on criteria other than the issuer claim? If it doesn’t find any matching criteria in the first non-deterministically chosen config, does it keep checking through any others until all options are exhausted? Or does the request fail somehow?

Could you provide some clarity on this?

**Teddy Torres [10:34 PM]**
Yes, it is non-deterministic (as it's always been) since we cannot guarantee the order in which config admin will order these configs. It might not be as it appears in the server.xml. We are just stating what it always has been.

**Teddy Torres[10:41 PM]**
Please see my comments also in https://github.com/OpenLiberty/blogs/pull/2575

Hi. 
Yes, it is non-deterministic (as it's always been prior to this change) since we cannot guarantee the order in which config admin will order these configs. It might not be as it appears in the server.xml. We are just stating what it always has been.
Member
@teddyjtorres teddyjtorres 1 minute ago
Therefore, to avoid confusion, the following is proposed, since this is what it always has done,

If the issuer claim from the JWT does not match the issuer attribute of any openidConnectClient configuration elements, choose the first configuration that matched.

**Teddy Torres [11:11 PM]**
Hi Ramkumar, please leave the description as it was
[11:11]
If the issuer claim from the JWT does not match the `issuer` attribute of any `openidConnectClient` configuration elements, choose one of the configuration elements in a nondeterministic manner.

**Comment by ramkumar-k-9286** [comment]
changes were added to vNext and published with the 22.0.0.10 release.

Closing this issue.

Thanks.
Ramkumar.
